### DISCUSSION TITLE
How to educate our developers about monorepos?

### DISCUSSION THREAD

**Discussion by KATT** [trpc/trpc#3507] [General]
> I feel like over half of the questions in [#🙋-help](https://discord.com/channels/867764511159091230/1020132255567454310) has to do with setup problems, esp monorepos. What can we do to help here? Troubleshooting guide? Blog posts?
>
> _- Originally posted [on Discord](https://discord.com/channels/867764511159091230/980118117676118067/1057345404947148890) by @KATT_

**Comment by c-ehrlich** [comment]
Copy pasting my comment from Discord (about both monorepos and the "getting started" docs experience in general):

sorry for the wall of text 😅
we have the same thing on ct3a side. i think the truth is that setting up tRPC is complicated. we're experimenting with better documentation ourselves (see https://github.com/t3-oss/create-t3-app/pull/934) because just like you, more than half of the questions we get are about setup stuff. Of course on tRPC side it's even more difficult because there are so many different setup possibilities, ct3a has to deal with just one.

A few thoughts:
- Several other projects like Next.js, React, etc. basically treat an init tool as the "official" way to set up an app. This is obviously more difficult for tRPC because of how many different ways there are to set it up, but some of these resources, sample repos, etc already exist so maybe their educational use can be enhanced?
- It might be possible to make the "Quickstart" and "Usage with Next.js" sections shorter / even more focused, which could help with beginner questions. Currently just these two articles alone add up to a ton of text/information to process
- Beyond that I think overall some docs refactoring would be possible (but this is very difficult in my experience). I just read through the Tanstack Query docs today, and of course they have a smaller API surface but I think they do a great job of splitting the intro/teaching bit from the api reference/kitchen sink bit.
- Why are https://trpc.io/docs/awesome-trpc#-starting-points-example-projects-etc and https://trpc.io/docs/example-apps different pages?
- Re: monorepo stuff
  - If you do a CMD+K for "monorepo" in the docs, not much shows up
  - In https://trpc.io/docs/awesome-trpc#-starting-points-example-projects-etc, maybe add "turborepo" and/or "monorepo" to the create-t3-turbo description to have it show up for that keyword?
  - A dedicated Monorepo page in the docs with some info for nx/turbo, a list of projects to reference (the nx starter, ct3-turbo, etc), and any other information that might be useful?

**Reply by guushamann** [reply]
in the discussion on discord we talked about the kind of monorepo's there are.

If we want to help people more we need to know there setup in my opinion.

a troubleshooting guide and a guide on how to ask for help if it involves a monorepo?

i agree with @cje maybe 1 or more best practice for monorepo guides would be awesome
i just created a monorepo setup with vue3, the main reason to do this with a very simple pet project was that the setup of the monorepo with all the bells and whistles is already a major task
for me it does not work to start a project from scratch, building a monorepo and at the same time already starting to work on what i want to build

for Npm workspaces,  a little list of things (tip of the iceberg) to check while building when you are new in monorepoland:
- did you put all the child repo's in a directory that is defined in the workspace section in the root package.json 
- if you did an npm install of a package in a child repo, rm-rf node_modules in root and run npm install in the root
- is you're node and npm version ok?
- do you have the same npm modules in different child repo's? are they the same version?
- if you suspect an npm package for a dependency problem, check which versions exist of a package in you're node modules folders with npm explain <name of package>

and in general when somebody asks a question about monorepo
- ask if they used a template or tool recommended on the trpc website, which one?
- do you have a good understanding of the monorepo methods you use? link to a list of good resources about them

**Comment by guushamann** [comment]
Other thought, eleborating on the awesome page. 
Begin with making an assessment of the templates and tools, find out which monorepo type they use and which package manager, for the tools maybe that is multiple package managers, i do not know.
After making this list you can better decide on what is the next step
